### ISSUE TITLE
Calling auto_call_delayed_io twice breaks delayed io

### ISSUE THREAD

**Issue by dwelch-r7** [dmendel/bindata#149]
With the following test everything passes as expected:

```ruby
require 'bindata'

# Uncommenting this line and having an extra auto_call_delayed_io breaks everything
# class OtherModelThatBreaksDnsInfo < BinData::Record
#   auto_call_delayed_io
# end

class String16 < BinData::String
  def assign(val)
    super(val.to_s.encode('utf-16le')).force_encoding('utf-16le')
  end

  def snapshot
    super.force_encoding('utf-16le')
  end

  def read_and_return_value(io)
    super.force_encoding('utf-16le')
  end
end

class Krb5UpnDnsInfo < BinData::Record
  auto_call_delayed_io
  endian :little
  uint16 :upn_length, value: -> { upn.num_bytes }
  uint16 :upn_offset
  uint16 :dns_domain_name_length, value: -> { dns_domain_name.num_bytes }
  uint16 :dns_domain_name_offset
  uint32 :flags
  delayed_io :upn, read_abs_offset: -> { self.abs_offset + upn_offset } do
    string16 read_length: :upn_length
  end
  delayed_io :dns_domain_name, read_abs_offset: -> { self.abs_offset + dns_domain_name_offset } do
    string16 read_length: :dns_domain_name_length
  end
end

data = "\x28\x00\x10\x00\x1e\x00\x38\x00\x01\x00\x00\x00\x00\x00\x00\x00" \
    "\x74\x00\x65\x00\x73\x00\x74\x00\x40\x00\x77\x00\x69\x00\x6e\x00" \
    "\x64\x00\x6f\x00\x6d\x00\x61\x00\x69\x00\x6e\x00\x2e\x00\x6c\x00" \
    "\x6f\x00\x63\x00\x61\x00\x6c\x00\x57\x00\x49\x00\x4e\x00\x44\x00" \
    "\x4f\x00\x4d\x00\x41\x00\x49\x00\x4e\x00\x2e\x00\x4c\x00\x4f\x00" \
    "\x43\x00\x41\x00\x4c\x00".b

result = Krb5UpnDnsInfo.read(data)
pp result

if result.upn != 'test@windomain.local'.encode('UTF-16LE')
  raise "upn failed reading"
else
  puts "upn parsed successfully"
end
```

Expected output:

```ruby
$ ruby bug.rb                                                                                                                                                                         ‹ruby-3.0.5@metasploit-framework›
{:upn_length=>40,
 :upn_offset=>16,
 :dns_domain_name_length=>30,
 :dns_domain_name_offset=>56,
 :flags=>1,
 :upn=>"test@windomain.local",
 :dns_domain_name=>"WINDOMAIN.LOCAL"}
upn parsed successfully
```

But uncommenting the placeholder model which calls `auto_call_delayed_io` , breaks the UPN being read successfully:

```diff
diff --git a/bug.rb b/bug.rb
index abd7861e7e..778ff510a5 100644
--- a/bug.rb
+++ b/bug.rb
@@ -1,9 +1,9 @@
 require 'bindata'
 
 # Uncommenting this line and having an extra auto_call_delayed_io breaks everything
-# class OtherModelThatBreaksDnsInfo < BinData::Record
-#   auto_call_delayed_io
-# end
+class OtherModelThatBreaksDnsInfo < BinData::Record
+  auto_call_delayed_io
+end
 
 class String16 < BinData::String
   def assign(val)
```

Now the result shows the model isn't being parsed correctly:
```
$ ruby bug.rb                                                                                                                                                                         ‹ruby-3.0.5@metasploit-framework›
{:upn_length=>0,
 :upn_offset=>16,
 :dns_domain_name_length=>0,
 :dns_domain_name_offset=>56,
 :flags=>1,
 :upn=>"",
 :dns_domain_name=>""}
bug.rb:49:in `<main>': upn failed reading (RuntimeError)
```

I believe this is caused by the guard clause here which stops the `auto_call_delayed_io` logic being wired up for the second model:

https://github.com/dmendel/bindata/blob/7ede3d1e6efb6da3be8ef076e402951d68030544/lib/bindata/delayed_io.rb#L162

**Comment by dwelch-r7** [comment]
Working around this issue by calling the mixin include directly instead of using the DSL helper:

```diff
diff --git a/bug.rb b/bug.rb
index abd7861e7e..228760a0bc 100644
--- a/bug.rb
+++ b/bug.rb
@@ -1,9 +1,9 @@
 require 'bindata'
 
 # Uncommenting this line and having an extra auto_call_delayed_io breaks everything
-# class OtherModelThatBreaksDnsInfo < BinData::Record
-#   auto_call_delayed_io
-# end
+class OtherModelThatBreaksDnsInfo < BinData::Record
+  auto_call_delayed_io
+end
 
 class String16 < BinData::String
   def assign(val)
@@ -20,7 +20,9 @@ class String16 < BinData::String
 end
 
 class Krb5UpnDnsInfo < BinData::Record
-  auto_call_delayed_io
+  include AutoCallDelayedIO
+
   endian :little
   uint16 :upn_length, value: -> { upn.num_bytes }
   uint16 :upn_offset
```


So potentially the fix for this might be:

```diff
      def auto_call_delayed_io
+       include AutoCallDelayedIO

        return if DelayedIO.method_defined? :initialize_instance_without_record_io       

-       include AutoCallDelayedIO
        DelayedIO.send(:alias_method, :initialize_instance_without_record_io, :initialize_instance)
        DelayedIO.send(:define_method, :initialize_instance) do
          if @parent && !defined? @delayed_io_recorded
            @delayed_io_recorded = true
            list = top_level_get(:delayed_ios)
            list << self if list
          end

          initialize_instance_without_record_io
        end
      end
```

**Comment by dmendel** [comment]
Fixed and a new release has been made (2.4.15).

Thanks for a very clear bug report.
